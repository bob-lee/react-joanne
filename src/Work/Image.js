import React from 'react'
import Loader from './Loader'

export default class Image extends React.Component {
  state = { loaded: false }

  loaded = () => {
    //console.log('loaded')
    this.setState({ loaded: true })
    this.props.onImageLoaded({
      element: this.imgElement,
      index: this.props.index
    })
  }

  render() {
    const item = this.props.item
    let classes = 'image'
    if (this.state.loaded) {
      classes += ' show'
    } 
    
    const style = { margin: '0 auto' }
    const loading = (item.toLoad && !this.state.loaded) ? <Loader style={style} /> : null

    let thumbImage = null
    const thumbUrl = item.toLoad && item.thumbUrl
    if (thumbUrl) {
      thumbImage = <img src={thumbUrl}
        className="image2"
        alt="thumb" />
    }

    return (
      <div>
        {loading}
        <div className={classes}>
          <picture>
            <img src={item.toLoad ? item.url : undefined}
              onLoad={this.loaded}
              className="image1"
              ref={(el) => { this.imgElement = el; }}
              alt={item.fileName} />
          </picture>

          <div className="expand">
            <span>{item.text}</span>
            {thumbImage}
          </div>
        </div>
      </div>
    )
  }
}
