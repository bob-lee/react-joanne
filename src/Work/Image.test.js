import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import toJson from 'enzyme-to-json'
import Image from './Image'

const getItem = (toLoad) => ({ 
  fileName: 'test.jpg',
  url: '/test.jpg',
  thumbUrl: '/test_thumb.jpg',
  text: 'test',
  toLoad: toLoad
})

const getProps = (toLoad) => ({
  item: getItem(toLoad),
  index: 0,
  onImageLoaded: jest.fn()/*({ element, index }) => {
    console.log(`onImageLoaded(${index}`, element)
  }*/
})

it('Image renders without crashing', () => {
  const props = getProps(true)
  const div = document.createElement('div')
  ReactDOM.render(<Image {...props} />, div)
})

it('Image renders correctly with toLoad=true', () => {
  const props = getProps(true)
  const tree = renderer
    .create(<Image {...props} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Image renders correctly with toLoad=false', () => {
  const props = getProps(false)
  const tree = renderer
    .create(<Image {...props} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it(`onImageLoaded should be called on 'load' event`, () => {
  const props = getProps(true)
  const wrapper = shallow(<Image {...props} />)
  expect(wrapper.state().loaded).toBe(false)
  wrapper.find('img.image1').simulate('load')
  console.log('mock.calls', props.onImageLoaded.mock.calls)
  expect(wrapper.state().loaded).toBe(true)
  expect(props.onImageLoaded.mock.calls.length).toBe(1)
  expect(props.onImageLoaded.mock.calls[0][0]).toEqual({element: undefined, index: props.index})
  expect(toJson(wrapper)).toMatchSnapshot()
})
