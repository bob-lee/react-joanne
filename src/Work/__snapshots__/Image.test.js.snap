// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders correctly with toLoad=false 1`] = `
<div>
  <div
    className="image"
  >
    <picture>
      <img
        alt="test.jpg"
        className="image1"
        onLoad={[Function]}
        src={undefined}
      />
    </picture>
    <div
      className="expand"
    >
      <span>
        test
      </span>
    </div>
  </div>
</div>
`;

exports[`Image renders correctly with toLoad=true 1`] = `
<div>
  <div
    className="loader"
    style={
      Object {
        "margin": "0 auto",
      }
    }
  >
    Loading...
  </div>
  <div
    className="image"
  >
    <picture>
      <img
        alt="test.jpg"
        className="image1"
        onLoad={[Function]}
        src="/test.jpg"
      />
    </picture>
    <div
      className="expand"
    >
      <span>
        test
      </span>
      <img
        alt="thumb"
        className="image2"
        src="/test_thumb.jpg"
      />
    </div>
  </div>
</div>
`;

exports[`onImageLoaded should be called on 'load' event 1`] = `
<div>
  <div
    className="image show"
  >
    <picture>
      <img
        alt="test.jpg"
        className="image1"
        onLoad={[Function]}
        src="/test.jpg"
      />
    </picture>
    <div
      className="expand"
    >
      <span>
        test
      </span>
      <img
        alt="thumb"
        className="image2"
        src="/test_thumb.jpg"
      />
    </div>
  </div>
</div>
`;
